Products/Zing/fact.py

```diff
@@ -171,33 +171,27 @@ def device_info_fact(device):
     f.metadata[DimensionKeys.PLUGIN_KEY] = DEVICE_INFO_FACT_PLUGIN
     f.data[MetadataKeys.NAME_KEY] = device.titleOrId()
     f.data[MetadataKeys.PROD_STATE_KEY] = device.getProductionStateString()
+    valid_types = (str, int, long, float, bool, list, tuple, set,)
     for propdict in device._propertyMap():
         propId = propdict.get("id")
-        if device.isLocal(propId) and propdict.get("type", None) in (
-            "string",
-            "int",
-            "boolean",
-            "long",
-            "float",
-            "text",
-            "lines",
-        ):
-            # Some of the device properties can be methods,
-            # so we have to call them and get values.
-            if callable(device.getProperty(propId)):
-                log.warn("Callable: %s", device.getProperty(propId))
-                try:
-                    value = device.getProperty(propId).__call__()
-                except TypeError as e:
-                    log.exception(
-                        "Unable to call property: %s. Exception %s",
-                        device.getProperty(propId),
-                        e,
-                    )
-            else:
-                value = device.getProperty(propId)
-            if value is None:
-                value = ""
+        value = None
+        # Some of the device properties can be methods,
+        # so we have to call them and get values.
+        if callable(device.getProperty(propId)):
+            log.warn("Callable: %s", device.getProperty(propId))
+            try:
+                value = device.getProperty(propId).__call__()
+            except TypeError as e:
+                log.exception(
+                    "Unable to call property: %s. Exception %s",
+                    device.getProperty(propId),
+                    e,
+                )
+        else:
+            value = device.getProperty(propId)
+        if value is None:
+            value = ""
+        if device.isLocal(propId) and isinstance(value, valid_types):
```

Comment:
You mention the original problem was triggered by "non-scalar" types.  Was that only dict, or will we have problems with list, tuple, or set?   Do we need to convert tuple and set to list, for instance, or is that handled elsewhere when the fact is processed?

Reply:
I believe the `lines` type that before assumed `list, set, tuple` as valid types for scalar arrays.

Reply:
Please see FactEncoder https://github.com/zenoss/zenoss-prodbin/blob/develop/Products/Zing/fact.py#L313-L338